### ISSUE TITLE
How to calculate epoch time for a dataset?

### ISSUE THREAD

**Issue by AnchorAnalytics** [AlexeyAB/darknet#6395]
I have trained several large datasets with Darknet with exceptable results.  My datasets are usually over  300K images.  I train on a laptop (batch = 64 and subdivision = 64 [@704x704]) so training takes a long time (2 to 3 days).   I have never trained for an entire epoch.  I stop training when loss is below 1 which in many cases (for the size of my datasets) is only 10 to 15% of the iterations required to complete an epoch.  When I train, I don't show/calculate mPA to save on computing resources.

I'm wondering if training for an entire epoch is worth the time or should I focus on dataset optimization.  I would like to calculate epoch time for my computer for a specific dataset.

Is there a formula or literature available to assist with calculating epoch time for datasets?

Thanks!

**Comment by Luux** [comment]
Batched Training scales linearly with the data set size. So if you know you need t hours to train i images, you can simply calculate t_total = i_total / i * t

Or did I misunderstand your problem?

**Comment by Yogerlan** [comment]
> I have trained several large datasets with Darknet with exceptable results. My datasets are usually over 300K images. I train on a laptop (batch = 64 and subdivision = 64 [@704x704]) so training takes a long time (2 to 3 days). I have never trained for an entire epoch. I stop training when loss is below 1 which in many cases (for the size of my datasets) is only 10 to 15% of the iterations required to complete an epoch. When I train, I don't show/calculate mPA to save on computing resources.
> 
> I'm wondering if training for an entire epoch is worth the time or should I focus on dataset optimization. I would like to calculate epoch time for my computer for a specific dataset.
> 
> Is there a formula or literature available to assist with calculating epoch time for datasets?
> 
> Thanks!

To improve detection is recommended to train for several epoch or your training will be worthless. I don't know your hardware specs but if you want to train with low resources, maybe you need to:

1. Reduce the dataset (small datasets compromises detection results, but can work in some specific cases).
2. Increase the training time.
3. Improve your hardware.

Others solutions are handy but can compromise your results.

**Comment by AnchorAnalytics** [comment]
@Luux @Yogerlan ,

I apologize for the delay and thank you for your reply.  My dataset has 12 classes and there is  25K images per class, a total of 300K images.   My config file parameters are below.  Note: max_batch = 24K.  However, during training, the loss chart drops below 1 (at approximately 1000 iterations) which is before the first epoch.  Based on my understanding of epoch:

1 epoch = #images / batch  ==> 300K / 64 = 4687 iterations.  The  Network has seen all the data once at 4687 iterations.

How can the loss be below 1 at 1000 iterations, when all the data has not passed through the Network?

Thanks!
 
Config File Settings:
[net]
batch=64
subdivisions=64
#Training
width=704
height=704
channels=3
momentum=0.949
decay=0.0005
angle=0
saturation = 1.5
exposure = 1.5
hue=.1

learning_rate=0.0013
burn_in=1000
max_batches=24000
policy=steps
steps=19200,21600
scales=.1,.1

**Comment by Yogerlan** [comment]
> @Luux @Yogerlan ,
> 
> I apologize for the delay and thank you for your reply. My dataset has 12 classes and there is 25K images per class, a total of 300K images. My config file parameters are below. Note: max_batch = 24K. However, during training, the loss chart drops below 1 (at approximately 1000 iterations) which is before the first epoch. Based on my understanding of epoch:
> 
> 1 epoch = #images / batch ==> 300K / 64 = 4687 iterations. The Network has seen all the data once at 4687 iterations.
> 
> How can the loss be below 1 at 1000 iterations, when all the data has not passed through the Network?
> 
> Thanks!
> 
> Config File Settings:
> [net]
> batch=64
> subdivisions=64
> #Training
> width=704
> height=704
> channels=3
> momentum=0.949
> decay=0.0005
> angle=0
> saturation = 1.5
> exposure = 1.5
> hue=.1
> 
> learning_rate=0.0013
> burn_in=1000
> max_batches=24000
> policy=steps
> steps=19200,21600
> scales=.1,.1

The Deep Learning experts say that increasing slowly the learning rate from zero can improve the starting accuracy (reduce the avg loss) of the model. That's why the `burn_in=1000` parameter is set. However, 1000 iterations doesn't complete an epoch, so the model needs more iterations to get a real accuracy estimation.

As i can see, you have adjusted the `max_batches` and `steps` parameters to the recommended values, the training should be fine with those. If you want to speed up the training, then try to reduce the network size (608x608 or 512x512) and the `subdivisions` 32, 24, 16...

**Comment by AnchorAnalytics** [comment]
@Yogerlan 
Thanks!  I did not know the burn_in adjusted during training.
Thanks again for your reply and time.  Issue closed.
